cpp/src/arrow/compute/exec/bloom_filter_avx2.cc

```diff
@@ -0,0 +1,136 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <immintrin.h>
+#include "arrow/compute/exec/bloom_filter.h"
+#include "arrow/util/bit_util.h"
+
+namespace arrow {
+namespace compute {
+
+#if defined(ARROW_HAVE_AVX2)
+
+inline __m256i BlockedBloomFilter::mask_avx2(__m256i hash) const {
+  // AVX2 translation of mask() method
+  //
+  __m256i mask_id =
+      _mm256_and_si256(hash, _mm256_set1_epi64x(BloomFilterMasks::kNumMasks - 1));
+
+  auto masks = reinterpret_cast<const arrow::util::int64_for_gather_t*>(masks_.masks_);
+  __m256i result = _mm256_i64gather_epi64(masks, _mm256_srli_epi64(mask_id, 3), 1);
+  result = _mm256_srlv_epi64(result, _mm256_and_si256(mask_id, _mm256_set1_epi64x(7)));
+  result = _mm256_and_si256(result, _mm256_set1_epi64x(BloomFilterMasks::kFullMask));
+
+  __m256i rotation = _mm256_and_si256(
+      _mm256_srli_epi64(hash, BloomFilterMasks::kLogNumMasks), _mm256_set1_epi64x(63));
+
+  result = _mm256_or_si256(
+      _mm256_sllv_epi64(result, rotation),
+      _mm256_srlv_epi64(result, _mm256_sub_epi64(_mm256_set1_epi64x(64), rotation)));
+
+  return result;
+}
+
+inline __m256i BlockedBloomFilter::block_id_avx2(__m256i hash) const {
```

Comment:
Is it worth to add `__attribute__((always_inline))` here?

Reply:
That is not supported by all compilers.